src/Sarif/EnumeratedArtifact.cs

```diff
@@ -16,44 +16,164 @@ public EnumeratedArtifact(IFileSystem fileSystem)
+        }
 
         internal IFileSystem FileSystem { get; set; }
 
+        /// <summary>
+        /// Gets or sets a property that determines whether the enumerated artifact
+        /// can obtain data from a non-seekable stream. This operation will be
+        /// less performant in any case where an artifact is a textual file (as
+        /// the file will first be converted to a byte array to determine
+        /// whether it is textual and subsequently turned into a string if it is).
+        /// </summary>
+        public bool SupportNonSeekableStreams { get; set; }
+
         public string Contents
         {
-            get => GetContents();
+            get => GetArtifactData().text;
             set => this.contents = value;
         }
 
-        private string GetContents()
+        public byte[] Bytes
+        {
+            get => GetArtifactData().bytes;
+            set => this.bytes = value;
+        }
+
+        private (string text, byte[] bytes) GetArtifactData()
         {
-            if (this.contents != null) { return this.contents; }
+            if (this.contents != null)
+            {
+                return (this.contents, bytes: null);
+            }
+
+            if (this.bytes != null)
+            {
+                return (text: null, this.bytes);
+            }
 
-            if (Stream == null && this.contents == null)
+            if (Stream == null && this.contents == null && this.bytes == null)
             {
-                // TBD we actually have no validation URI is non-null yet.
-                contents = Uri!.IsFile
-                    ? FileSystem.FileReadAllText(Uri.LocalPath)
-                    : null;
+                if (Uri == null ||
+                    !Uri.IsAbsoluteUri ||
+                    (Uri.IsAbsoluteUri && !Uri.IsFile))
+                {
+                    throw new InvalidOperationException("An absolute URI pointing to a file location was not available.");
+                }
 
-                this.sizeInBytes = (long?)this.contents?.Length;
+                // This is our client-side, disk-based file retrieval case.
+                Stream = FileSystem.FileOpenRead(Uri.LocalPath);
+            }
+
+            if (Stream.CanSeek)
+            {
+                RetrieveDataFromSeekableStream();
             }
             else
             {
-                if (Stream.CanSeek) { this.Stream.Seek(0, SeekOrigin.Begin); }
+                RetrieveDataFromNonSeekableStream();
+            }
+
+            Stream = null;
+
+            return (this.contents, this.bytes);
+        }
+
+        private void RetrieveDataFromNonSeekableStream()
+        {
+            bool isText;
+            if (!SupportNonSeekableStreams)
+            {
+                throw new InvalidOperationException("Stream is not seekable. Provide a seekable stream or set the 'SupportNonSeekableStreams' property.");
+            }
+
+            this.bytes = new byte[Stream.Length];
```

Comment:
>bytes [](http://example.com/codeflow?start=17&length=5)

should we consider adding lock to the bytes array in case a thread tries to read dirty data?